x/liquidstaking/keeper/derivative.go

```diff
@@ -0,0 +1,163 @@
+)
+
+// MintDerivative mints a new derivative
+func (k Keeper) MintDerivative(ctx sdk.Context, delegatorAddr sdk.AccAddress, valAddr sdk.ValAddress, coin sdk.Coin) error {
+
+	_, found := k.stakingKeeper.GetValidator(ctx, valAddr)
+	if !found {
+		return types.ErrNoValidatorFound
+	}
+
+	validator, found := k.stakingKeeper.GetValidator(ctx, valAddr)
+	if !found {
+		return types.ErrNoValidatorFound
+	}
+
+	delegation, found := k.stakingKeeper.GetDelegation(ctx, delegatorAddr, valAddr)
+	if !found {
+		return types.ErrNoDelegatorForAddress
+	}
+
+	if coin.Denom != k.stakingKeeper.BondDenom(ctx) {
+		return types.ErrOnlyBondDenomAllowdForTokenize
+	}
+
+	delegationAmount := validator.Tokens.ToDec().Mul(delegation.GetShares()).Quo(validator.DelegatorShares)
+	if coin.Amount.GT(sdk.Int(delegationAmount)) {
+		return types.ErrNotEnoughDelegationShares
+	}
+
+	acc := k.accountKeeper.GetAccount(ctx, delegatorAddr)
+	if acc != nil {
+		acc, ok := acc.(vesting.VestingAccount)
+		if ok {
+			// if account is a vesting account, it checks if free delegation (non-vesting delegation) is not exceeding
+			// the tokenize share amount and execute further tokenize share process
+			// tokenize share is reducing unlocked tokens delegation from the vesting account and further process
+			// is not causing issues
+			delFree := acc.GetDelegatedFree().AmountOf(coin.Denom)
+			if delFree.LT(coin.Amount) {
+				return types.ErrExceedingFreeVestingDelegations
+			}
+		}
+	}
+
+	liquidCoinDenom := k.GetLiquidStakingTokenDenom(ctx, valAddr)
+	liquidCoin := sdk.NewCoin(liquidCoinDenom, coin.Amount)
+
+	delegationHolder, found := k.GetDelegationHolder(ctx, valAddr)
+	if !found {
+		// Generate specific module account address for this valAddr
+		modAccName := authtypes.NewModuleAddress(fmt.Sprintf("%s:%s", types.ModuleName, valAddr.String()))
+		modBaseAcc := authtypes.NewBaseAccount(modAccName, nil, 0, 0)
+		modAcc := authtypes.NewModuleAccount(modBaseAcc, types.ModuleName, []string{}...)
+		k.accountKeeper.SetModuleAccount(ctx, modAcc)
+
+		delegationHolder = types.NewDelegationHolder(modAcc.GetAddress(), valAddr, liquidCoin)
+	} else {
+		// Check that the existing delegation liquid staking token denom matches
+		if delegationHolder.Delegations.Denom != liquidCoinDenom {
+			return types.ErrInvalidLiquidCoinDenom
+		}
+		// Add the liquid staking coin to the delegations
+		delegationHolder.Delegations = delegationHolder.Delegations.Add(liquidCoin)
+	}
+
+	err := k.bankKeeper.MintCoins(ctx, minttypes.ModuleName, sdk.Coins{liquidCoin})
+	if err != nil {
+		return err
+	}
+
+	err = k.bankKeeper.SendCoinsFromModuleToAccount(ctx, minttypes.ModuleName, delegatorAddr, sdk.Coins{liquidCoin})
+	if err != nil {
+		return err
+	}
+
+	shares, err := k.stakingKeeper.ValidateUnbondAmount(
+		ctx, delegatorAddr, valAddr, coin.Amount,
+	)
+
+	returnAmount, err := k.stakingKeeper.Unbond(ctx, delegatorAddr, valAddr, shares)
+	if err != nil {
+		return err
+	}
```

Comment:
I think it'll be easier to keep the derivative amount equivalent to the delegator shares. eg provide the number of shares you want to convert to derivative rather than the amount of ukava, then mint derivatives equal to that number.

Though we'll need to convert shares from `sdk.Dec` to `Int` somewhere which could complicate things.

Reply:
https://github.com/Kava-Labs/kava/pull/1271/commits/200e9785c49f4522180545dba6744b42d3691f72